### PR TITLE
auto-chunk unembed & loss

### PR DESCRIPTION
**Pull request by shunting314**
Add the ability to compile the loss together with the unembed linear layer. The benefit is that we would be able to chunk the logits (which is usually quite large due to large vocab size) by the compiler and reduce peak memory usage.

Here are testing results on qwen3 1.7B. With batch-size=16, the baseline uses 115.85GiB peak memory and gets 54_450 tps. 
By applying the autochunker, we uses 84.15GiB peak memory with 54_244 tps. This is ***37.7%*** peak memory saving trading with ***0.38%*** perf loss. For larger model, the percentage of saving can be smaller since memory usage per layer and activations/optimizer state becomes larger. But saving peak memory usage by auto-chunking is still very nice if perf trade off is very small.

Command for baseline: ```NGPU=1 CONFIG_FILE=torchtitan/models/qwen3/train_configs/qwen3_1.7b.toml ./run_train.sh --compile.enable --training.local_batch_size=16 ```
Command enabling autochunking: ```NGPU=1 CONFIG_FILE=torchtitan/models/qwen3/train_configs/qwen3_1.7b.toml ./run_train.sh --compile.enable --training.local_batch_size=16 --compile.components=model,unembed_and_loss ```

To enable a model for auto-chunking, one tiny change is needed. The forward method need to have an 'unembed' boolean arguments. If it's false, the forward method should not do the unembed linear compuation so we can do that together with the loss compulation

cc @jansel , @eellison , @v0i0

### CODE REVIEW

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Agreed that chunked loss computation is an important feature. It's very nice that we can do it "automatically" with compile!

However, there are several worries:
1. Does it work with FSDP? IIUC, putting model.output in the compile region would cause graph break. Is it not the case?
2. Does it work with loss parallel on the TP mesh? This can be verified by setting `tensor_parallel_degree` > 1.
3. The change is intrusive and making things hard to reason. I guess one "proper" way of doing it might be introducing an output_processor / logits_processor in transformer model code (which also aligns with libraries like vllm) and apply compile to them in parallelize.py.
4. Besides, maybe it's worth adding the eager chunked loss as baseline.

**Comment by shunting314** [comment]
Thanks for reviewing @tianyu-l .

> Besides, maybe it's worth adding the eager chunked loss as baseline.

Is there any specific eager chunked loss you are looking for? This one https://github.com/linkedin/Liger-Kernel/blob/6a383424208b1d79bca2462f7d93bcfb9d13da05/src/liger_kernel/ops/fused_linear_cross_entropy.py#L279 ?

**Comment by tianyu-l** [comment]
oh, I don't think we have to define backward?
e.g. you can check https://fburl.com/code/mfx2xodx

File: torchtitan/models/qwen3/model/model.py [hunk 1]
```diff
@@ -586,5 +587,5 @@ def forward(
         # pyrefly: ignore [not-callable]
         h = self.norm(h) if self.norm else h
```
Comment (wwwjn) [inline torchtitan/models/qwen3/model/model.py:588]: Thank you! I have a n00b question - why we don't compile this norm together and let the `self.norm` get chunked as well? 

I was asking because I see vLLM / other model usual split the forward pass into 2 functions (as @tianyu-l  suggested):
1). Transfomers layers forward
2). post transformer layer process 

`self.norm` sometimes is put in 1) (Eg, vllm logit_processor), but sometimes it was put in 2).

**Comment by tianyu-l** [comment]
Is it right that `loss.backward` call doesn't need to be in the compile region? I think we might need to put loss function into model code. cc @wwwjn

File: torchtitan/models/qwen3/model/model.py (hunk 1 above)
Comment (shunting314) [reply torchtitan/models/qwen3/model/model.py:588]: the `self.norm` is not a good candidate for chunking. A simplified answer is, ops operating on tensors with a V (vocabulary size) dimension get chunked. The tensors with a V dimension are quite large since V is large and chunking them would be beneficial to reduce peak memory. But even if we change the model implementation to compile `self.norm` together with linear+loss, Inductor would still be able to skip `self.norm` for chunking

Comment (wwwjn) [reply torchtitan/models/qwen3/model/model.py:588]: Thank you, chunking on Vocab_size dimension makes sense!

**Comment by wwwjn** [comment]
> Is it right that `loss.backward` call doesn't need to be in the compile region? I think we might need to put loss function into model code. cc @wwwjn

Let me try to understand this more: So if we put loss function into model code, we could compile self.output() + loss forward  together as a single region?

**Comment by tianyu-l** [comment]
> Let me try to understand this more: So if we put loss function into model code, we could compile self.output() + loss forward together as a single region?

@wwwjn I think so?

File: torchtitan/models/qwen3/model/model.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/qwen3/model/model.py:588]: @shunting314 sorry I still have some confusion

According to your description, the auto-chunk is similar to https://github.com/apple/ml-cross-entropy and chunks on vocab dimension.

In contrast, https://fburl.com/code/mfx2xodx chunks on the sequence dimension, whose loss computation is simpler because there's no aggregation across chunks.

May I ask what's the benefit of chunking on vocab dim compared with batch / sequence dim?

Comment (shunting314) [reply torchtitan/models/qwen3/model/model.py:588]: > According to your description, the auto-chunk is similar to https://github.com/apple/ml-cross-entropy and chunks on vocab dimension.

AutoChunker still chunks on the 'flattened' batch+seqlen dimension. I mention vocab-size above since that's a big motivation to chunk the tensors/ops. If the tensor is small, chunking does not bring much benefit.
